core/src/main/scala/org/apache/spark/util/collection/OpenHashSet.scala

```diff
@@ -118,8 +125,21 @@ class OpenHashSet[@specialized(Long, Int, Double, Float) T: ClassTag](
    * See: https://issues.apache.org/jira/browse/SPARK-45599
    */
   @annotation.nowarn("cat=other-non-cooperative-equals")
-  private def keyExistsAtPos(k: T, pos: Int) =
-    _data(pos) equals k
+  protected def keyExistsAtPos(k: T, pos: Int) = {
+    classTag[T] match {
+      case ClassTag.Long => _data(pos) equals k
+      case ClassTag.Int => _data(pos) equals k
+      case ClassTag.Double => _data(pos) equals k
+      case ClassTag.Float => _data(pos) equals k
+      case _ => nonClassTagKeyExistsAtPos(k, _data(pos))
```

Comment:
wait, since this is `specialized(Long, Int, Double, Float)`, what else could be here in `case _`?

Reply:
@uros-db the specialized annotation is a performance optimization for versions of the generic class with those primitive types. But the type can still be anything else. The annotation results in the compiler generating the versions of generic classes for the specific types being specialized. The generic class can still be any other type, but boxing and unboxing will occur. 

Source: https://www.scala-lang.org/api/current/scala/specialized.html
 https://www.waitingforcode.com/scala-types/type-specialization-scala/read#:~:text=Scala%20uses%20the%20%40specialized%20class,classes%20for%20the%20specific%20types.
https://www.baeldung.com/scala/specialized-annotation